packages/gatsby/src/schema/extensions/index.js

```diff
@@ -0,0 +1,210 @@
+          `"dateformat", "link" and "fileByRelativePath".`,
+      },
+      options: {
+        type: GraphQLJSON,
+        description: `Resolver options. Vary based on resolver type.`,
+      },
+    },
+    process(args, fieldConfig) {
+      const { process } = fieldExtensions[args.type] || {}
+      if (typeof process === `function`) {
+        return process(args.options || {}, fieldConfig)
+      }
+      return {}
+    },
+  },
+
+  dateformat: {
+    description: `Add date formating options.`,
+    args: {
+      formatString: { type: GraphQLString },
+      locale: { type: GraphQLString },
+    },
+    process(args, fieldConfig) {
+      return getDateResolver(args)
+    },
+  },
+
+  link: {
+    description: `Link to node by foreign-key relation.`,
+    args: {
+      by: {
+        type: new GraphQLNonNull(GraphQLString),
+        defaultValue: `id`,
+      },
+      from: {
+        type: GraphQLString,
+      },
+    },
+    process(args, fieldConfig) {
+      return {
+        resolve: link(args),
+      }
+    },
+  },
+
+  fileByRelativePath: {
+    description: `Link to File node by relative path.`,
+    args: {
+      from: {
+        type: GraphQLString,
+      },
+    },
+    process(args, fieldConfig) {
+      return {
+        resolve: fileByPath(args),
+      }
+    },
+  },
+
+  // projection: {
+  //   description: `Automatically add fields to selection set.`,
+  //   args: {},
+  //   process(args, fieldConfig) {},
+  // },
+
+  proxyFrom: {
+    description: `Proxy resolver from another field.`,
+    process(from, fieldConfig) {
+      const resolver = fieldConfig.resolve || defaultFieldResolver
+      return {
+        resolve(source, args, context, info) {
+          return resolver(source, args, context, {
+            ...info,
+            fieldName: from,
+          })
+        },
+      }
+    },
+  },
+}
+
+const toDirectives = ({ extensions, locations }) =>
+  Object.keys(extensions).map(name => {
+    const extension = extensions[name]
+    const { args, description } = extension
+    return new GraphQLDirective({ name, args, description, locations })
+  })
+
+const addDirectives = ({ schemaComposer }) => {
+  const fieldDirectives = toDirectives({
+    extensions: fieldExtensions,
+    locations: [DirectiveLocation.FIELD_DEFINITION],
+  })
+  fieldDirectives.forEach(directive => schemaComposer.addDirective(directive))
+  const typeDirectives = toDirectives({
+    extensions: typeExtensions,
+    locations: [DirectiveLocation.OBJECT],
+  })
+  typeDirectives.forEach(directive => schemaComposer.addDirective(directive))
+}
+
+const processFieldExtensions = ({
+  schemaComposer,
+  typeComposer,
+  parentSpan,
+}) => {
+  typeComposer.getFieldNames().forEach(fieldName => {
+    const extensions = typeComposer.getFieldExtensions(fieldName)
+    Object.keys(extensions)
+      .filter(name => !internalExtensionNames.includes(name))
+      .sort(a => a === `proxyFrom`) // Ensure `proxyFrom` is run last
```

Comment:
I don't like those special cases. Now it feels extensions should be a list.

Reply:
won't you have the same issue with a list?